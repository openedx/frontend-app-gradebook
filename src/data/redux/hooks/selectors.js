import { useSelector } from 'react-redux';

import { StrictDict } from 'utils';
import selectors from 'data/selectors';

const selectorHook = (selector) => () => useSelector(selector);

export const root = StrictDict({
  useEditModalPossibleGrade: selectorHook(selectors.root.editModalPossibleGrade),
  useGetHeadings: selectorHook(selectors.root.getHeadings),
  useGradeExportUrl: selectorHook(selectors.root.gradeExportUrl),
  useInterventionExportUrl: selectorHook(selectors.root.interventionExportUrl),
  useSelectedCohortEntry: selectorHook(selectors.root.selectedCohortEntry),
  useSelectedTrackEntry: selectorHook(selectors.root.selectedTrackEntry),
  useShouldShowSpinner: selectorHook(selectors.root.shouldShowSpinner),
  useShowBulkManagement: selectorHook(selectors.root.showBulkManagement),
  useFilterBadgeConfig: (filterName) => useSelector(
    (state) => selectors.root.filterBadgeConfig(state, filterName),
  ),
});

export const app = StrictDict({
  useActiveView: selectorHook(selectors.app.activeView),
  useAssignmentGradeLimits: selectorHook(selectors.app.assignmentGradeLimits),
  useAreCourseGradeFiltersValid: selectorHook(selectors.app.areCourseGradeFiltersValid),
  useCourseGradeLimits: selectorHook(selectors.app.courseGradeLimits),
  useCourseGradeFilterValidity: selectorHook(selectors.app.courseGradeFilterValidity),
  useCourseId: selectorHook(selectors.app.courseId),
  useModalData: selectorHook(selectors.app.modalData),
  useSearchValue: selectorHook(selectors.app.searchValue),
  useShowImportSuccessToast: selectorHook(selectors.app.showImportSuccessToast),
});

export const assignmentTypes = StrictDict({
  useAllAssignmentTypes: selectorHook(selectors.assignmentTypes.allAssignmentTypes),
  useAreGradesFrozen: selectorHook(selectors.assignmentTypes.areGradesFrozen),
});

export const cohorts = StrictDict({
  useAllCohorts: selectorHook(selectors.cohorts.allCohorts),
  // maybe not needed?
  useCohortsByName: selectorHook(selectors.cohorts.cohortsByName),
});

export const filters = StrictDict({
  useData: selectorHook(selectors.filters.allFilters),
  useIncludeCourseRoleMembers: selectorHook(selectors.filters.includeCourseRoleMembers),
  useSelectableAssignmentLabels: selectorHook(selectors.filters.selectableAssignmentLabels),
  useSelectedAssignmentLabel: selectorHook(selectors.filters.selectedAssignmentLabel),
  useAssignmentType: selectorHook(selectors.filters.assignmentType),
});

export const grades = StrictDict({
  useAllGrades: selectorHook(selectors.grades.allGrades),
  useUserCounts: () => ({
    filteredUsersCount: useSelector(selectors.grades.filteredUsersCount),
    totalUsersCount: useSelector(selectors.grades.totalUsersCount),
  }),
  useGradeData: selectorHook(selectors.grades.gradeData),
  useHasOverrideErrors: selectorHook(selectors.grades.hasOverrideErrors),
  useShowSuccess: selectorHook(selectors.grades.showSuccess),
  useSubsectionGrade: ({ gradeFormat, subsection }) => () => (
    selectors.grades.subsectionGrade[gradeFormat](subsection)
  ),
});

export const roles = StrictDict({
  useCanUserViewGradebook: selectorHook(selectors.roles.canUserViewGradebook),
});

export const tracks = StrictDict({
  useAllTracks: selectorHook(selectors.tracks.allTracks),
  // maybe not needed?
  useTracksByName: selectorHook(selectors.tracks.tracksByName),
});

export default StrictDict({
  app,
  assignmentTypes,
  cohorts,
  filters,
  grades,
  roles,
  tracks,
  root,
});
