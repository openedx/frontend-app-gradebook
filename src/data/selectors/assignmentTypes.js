import { StrictDict } from '@src/utils';

const areGradesFrozen = ({ assignmentTypes }) => assignmentTypes.areGradesFrozen;
const allAssignmentTypes = ({ assignmentTypes }) => assignmentTypes.results;

export default StrictDict({
  areGradesFrozen,
  allAssignmentTypes,
});
