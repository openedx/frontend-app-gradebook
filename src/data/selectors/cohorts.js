import { StrictDict } from '../../utils';
import * as module from './cohorts';

/**
 * allCohorts(state)
 * returns top-level cohorts results data
 * @param {object} state - redux state
 * @return {object[]} - list of cohort objects from fetch results
 */
export const allCohorts = (state) => state.cohorts.results || [];

/**
 * getCohortById(state, selectedCohortId)
 * returns cohort with given id
 * @param {object} state - redux state
 * @param {number} selectedCohortId - id of cohort to return
 * @return {object} cohort with given id.
 */
export const getCohortById = (state, selectedCohortId) => {
  const cohort = module.allCohorts(state).find(({ id }) => id === selectedCohortId);
  return cohort;
};

/**
 * getCohortNameById(state, selectedCohortId)
 * return name of cohort with given id
 * @param {object} state - redux state
 * @param {number} selectedCohortId - id of cohort whose name is requested
 * @return {string} - name of cohort with the given id
 */
const getCohortNameById = (state, selectedCohortId) => (
  module.getCohortById(state, selectedCohortId) || {}
).name;

/**
 * cohortsByName(state)
 * returns an object of all cohorts keyed by name
 * @param {object} state - redux state
 * @return {object} - all cohorts, keyed by name
 */
const cohortsByName = (state) => module.allCohorts(state).reduce(
  (obj, cohort) => ({ ...obj, [cohort.name]: cohort }),
  {},
);

export default StrictDict({
  cohortsByName,
  getCohortById,
  getCohortNameById,
  allCohorts,
});
