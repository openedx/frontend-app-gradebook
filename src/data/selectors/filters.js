/* eslint-disable import/no-self-import */
import { StrictDict } from 'utils';

import initialFilters from 'data/constants/filters';
import simpleSelectorFactory from '../utils';
import * as module from './filters';

// Transformers
/**
 * isDefault(name, value)
 * returns true iff the value is equal to the initial filter value
 * associated with the given name
 * @param {string} name - api filter name
 * @param {string/number/bool} value - filter value
 * @return {bool} - is this the default value for the given filter?
 */
export const isDefault = (name, value) => (
  value === initialFilters[name]
);

/**
 * chooseRelevantAssignmentData(assignment)
 * formats the assignment api data for an assignment object for consumption
 * @param {object} assignment - assignment data to prepare
 * @return {object} - formatted data ({ label, subsectionLabel, type, id })
 */
export const chooseRelevantAssignmentData = ({
  label,
  subsection_name: subsectionLabel,
  category: type,
  module_id: id,
}) => ({
  label, subsectionLabel, type, id,
});

/**
 * getAssignmentsFromResultsSubstate(results)
 * returns the section_breakdown of the first results entry
 * defaulting to an empty list.
 * @param {[object[]]} results - list of result entries from grades fetch
 * @return {object} - section_breakdown of first grade entry
 */
export const getAssignmentsFromResultsSubstate = (results) => (
  (results[0] || {}).section_breakdown || []
);

/**
 * relevantAssignmentDataFromResults
 * returns assignment info from grades results for the assignment with the given id
 * @param {object} grades - grades fetch result
 * @param {string} id - selected assignment id from assignment filter
 * @return {object} assignment data with type, label, and subsectionLabel
 */
export const relevantAssignmentDataFromResults = (grades, id) => (
  module.getAssignmentsFromResultsSubstate(grades)
    .map(module.chooseRelevantAssignmentData)
    .find(assignment => assignment.id === id)
);

// Selectors
/**
 * allFilters(state)
 * returns the top-level filter state.
 * @param {object} state - redux state
 * @return {object} - filter substate from redux state
 */
export const allFilters = (state) => state.filters || {};

/**
 * areAssignmentGradeFiltersSet(state)
 * Returns true iff either assignmentGradeMax or assignmentGradeMax is set.
 * @param {object} state - redux state
 * @return {bool} - are assignmentGrade filters set?
 */
export const areAssignmentGradeFiltersSet = ({ filters }) => !!(
  filters.assignmentGradeMax || filters.assignmentGradeMin
);

/**
 * selectableAssignments(state)
 * @param {object} state - redux state
 * @return {object[]} - list of selectable assignment objects, filtered if there is an
 *   assignmentType filter selected.
 */
export const selectableAssignments = (state) => {
  const selectedAssignmentType = allFilters(state).assignmentType;
  const needToFilter = selectedAssignmentType && selectedAssignmentType !== 'All';
  const allAssignments = getAssignmentsFromResultsSubstate(state.grades.results);
  if (needToFilter) {
    return allAssignments.filter(
      (assignment) => assignment.category === selectedAssignmentType,
    );
  }
  return allAssignments;
};

/**
 * Returns the relevant assignment data for all selectable assignments
 * @param {object} state - redux state
 * @return {object[]} - object of assignment data entries [({ label, subsectionLabel, type, id })]
 */
export const selectableAssignmentLabels = (state) => (
  selectableAssignments(state).map(chooseRelevantAssignmentData)
);

export const simpleSelectors = simpleSelectorFactory(
  ({ filters }) => filters,
  [
    'assignment',
    'assignmentGradeMax',
    'assignmentGradeMin',
    'assignmentType',
    'cohort',
    'courseGradeMax',
    'courseGradeMin',
    'track',
    'includeCourseRoleMembers',
  ],
);
/**
 * Returns the id of the selected assignment filter
 * @param {object} state - redux state
 * @return {string} - assignment id
 */
export const selectedAssignmentId = (state) => (simpleSelectors.assignment(state) || {}).id;

/**
 * selectedAssignmentLabel(state)
 * Returns the label of the selected assignment filter
 * @param {object} state - redux state
 * @return {string} - assignment label
 */
export const selectedAssignmentLabel = (state) => (simpleSelectors.assignment(state) || {}).label;

/**
 * Returns the api value for excludedCourseRoles based on the
 * internal Bool value for includeCourseRoleMembers.
 * @param {object} state - redux state
 * @return {string} - '' if to be included, else 'all'
 */
export const excludedCourseRoles = (state) => (
  simpleSelectors.includeCourseRoleMembers(state) ? '' : 'all'
);

export default StrictDict({
  ...simpleSelectors,
  isDefault,
  relevantAssignmentDataFromResults,
  selectedAssignmentId,
  selectedAssignmentLabel,
  selectableAssignmentLabels,
  selectableAssignments,
  allFilters,
  areAssignmentGradeFiltersSet,
  chooseRelevantAssignmentData,
  excludedCourseRoles,
  getAssignmentsFromResultsSubstate,
});
