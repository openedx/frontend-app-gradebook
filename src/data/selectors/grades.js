/* eslint-disable import/no-self-import */
import { StrictDict } from 'utils';

import { Headings, GradeFormats } from 'data/constants/grades';
import { formatDateForDisplay } from 'data/actions/utils';
import { getLocale, isRtl } from '@edx/frontend-platform/i18n';
import simpleSelectorFactory from '../utils';
import * as module from './grades';

/**
 * getRowsProcessed(historyEntryData)
 * @param {object} historyEntryData - the data param from a bulk management history entry
 * @return {string} - user-facing summary of total/skipped/failed rows.
 */
export const getRowsProcessed = ({
  processed_rows: processed,
  saved_rows: saved,
  total_rows: total,
}) => {
  const failed = processed - saved;
  const skipped = total - processed;
  const summaryEntries = [`${total} Students: ${saved} processed`];
  if (skipped > 0) { summaryEntries.push(`${skipped} skipped`); }
  if (failed > 0) { summaryEntries.push(`${failed} failed`); }
  return summaryEntries.join(', ');
};

/**
 * formatGradeOverrideForDisplay(historyArray)
 * returns the grade override history results in display format.
 * @param {object[]} historyArray - array of gradeOverrideHistoryResults
 * @return {object[]} - display-formatted history results ({ date, grader, reason, adjustedGrade })
 */
export const formatGradeOverrideForDisplay = historyArray => historyArray.map(item => ({
  date: formatDateForDisplay(new Date(item.history_date)),
  grader: item.history_user,
  reason: item.override_reason,
  adjustedGrade: item.earned_graded_override,
}));

export const minGrade = '0';
export const maxGrade = '100';

/**
 * formatMaxCourseGrade(percentGrade)
 * Takes a percent grade and returns it unless it is equal to the max grade
 * @param {string} percentGrade - grade percentage
 * @return {string} percent grade or null
 */
export const formatMaxCourseGrade = (percentGrade) => (
  (percentGrade === maxGrade) ? null : percentGrade
);
/**
 * formatMinCourseGrade(percentGrade)
 * Takes a percent grade and returns it unless it is equal to the min grade
 * @param {string} percentGrade - grade percentage
 * @return {string} percent grade or null
 */
export const formatMinCourseGrade = (percentGrade) => (
  (percentGrade === minGrade) ? null : percentGrade
);

/**
 * formatMaxAssignmentGrade(percentGrade, options)
 * Takes a percent grade and returns it unless it is equal to the max grade or
 * the assignment id is set.
 * @param {string} percentGrade - grade percentage
 * @param {object} options - options object ({ assignmentId });
 * @return {string} percent grade or null
 */
export const formatMaxAssignmentGrade = (percentGrade, options) => (
  (percentGrade === maxGrade || !options.assignmentId) ? null : percentGrade
);

/**
 * formatMinAssignmentGrade(percentGrade, options)
 * Takes a percent grade and returns it unless it is equal to the min grade or
 * the assignment id is set.
 * @param {string} percentGrade - grade percentage
 * @param {object} options - options object ({ assignmentId });
 * @return {string} percent grade or null
 */
export const formatMinAssignmentGrade = (percentGrade, options) => (
  (percentGrade === minGrade || !options.assignmentId) ? null : percentGrade
);

/**
 * headingMapper(category, label='All')
 * Takes category and label filters and returns a method that will take a section breakdown
 * and return the appropriate table headings.
 * @param {string} category - assignment filter type
 * @param {string} label - assignment filter label
 * @return {string[]} - list of table headers
 */
export const headingMapper = (category, label = 'All') => {
  const filters = {
    all: section => section.label,
    byCategory: section => section.label && section.category === category,
    byLabel: section => section.label && section.label === label,
  };

  let filter;
  if (label === 'All') {
    filter = category === 'All' ? filters.all : filters.byCategory;
  } else {
    filter = filters.byLabel;
  }
  const { username, email, totalGrade } = Headings;
  const filteredLabels = (entry) => entry.filter(filter).map(s => s.label);

  return (entry) => (
    entry
      ? [username, email, ...filteredLabels(entry), totalGrade]
      : []
  );
};

/**
 * transformHistoryEntry(rawEntry)
 * Takes a raw bulkManagementHistory entry and formats it for consumption
 * @param {object} rawEntry - raw history entry to transform
 * @return {object} - transformed history entry
 *  ({ timeUploaded, originalFilename, resultsSummary, ... })
 */
export const transformHistoryEntry = ({
  modified,
  original_filename: originalFilename,
  data,
  unique_id: courseId,
  id,
  ...rest
}) => ({
  timeUploaded: formatDateForDisplay(new Date(modified)),
  originalFilename,
  resultsSummary: {
    rowId: id,
    text: module.getRowsProcessed(data),
  },
  ...rest,
});

/**
 * roundGrade(val)
 * Takes a number and rounds it to 2 decimal places
 * defaults to 0
 * @param {number=0} val - value to round.
 * @return {number} - rounded number
 */
export const roundGrade = val => parseFloat((val || 0).toFixed(2));
export const subsectionGrade = StrictDict({
  /**
   * subsectionGrade.absolute(subsection)
   * returns rounded {earned}/{possible} if attempted, else ${earned}
   * @param {object} subsection - grade subsection entry
   * @return {string} - absolute-formatted subsection grade string
   */
  [GradeFormats.absolute]: (subsection) => {
    const earned = module.roundGrade(subsection.score_earned);
    const possible = module.roundGrade(subsection.score_possible);
    return subsection.attempted ? `${earned}${isRtl(getLocale()) ? '\\' : '/'}${possible}` : `${earned}`;
  },
  /**
   * subsectionGrade.percent(subsection)
   * returns rounded percent times 100
   * @param {object} subsection - grade subsection entry
   * @return {string} - percent-formatted subsection grade string
   */
  [GradeFormats.percent]: (subsection) => (
    module.roundGrade(subsection.percent * 100)
  ),
});

// Selectors
/**
 * allGrades(state)
 * returns the top-level redux grades state.
 * @param {object} state - redux state
 * @return {object} - redux grades state
 */
export const allGrades = ({ grades: { results } }) => results;

/**
 * bulkImportError(state)
 * returns the stringified bulkManagement import error messages.
 * @param {object} state - redux state
 * @return {string} - bulk import error messages joined into a display form
 *   (or empty string if there are none)
 */
export const bulkImportError = ({ grades: { bulkManagement } }) => (
  (!!bulkManagement && bulkManagement.errorMessages)
    ? `Errors while processing: ${bulkManagement.errorMessages.join('; ')};`
    : ''
);

/**
 * bulkManagementHistory(state)
 * returns the bulkManagement history entries from the grades state
 * @param {object} state - redux state
 * @return {object[]} - list of bulkManagement history entries
 */
export const bulkManagementHistory = ({ grades: { bulkManagement } }) => (
  (bulkManagement.history || [])
);

/**
 * bulkManagementHistoryEntries(state)
 * returns transformed history entries from bulkManagement grades data.
 * @param {object} state - redux state
 * @return {object[]} - list of transformed bulkManagement history entries
 */
export const bulkManagementHistoryEntries = (state) => (
  module.bulkManagementHistory(state).map(module.transformHistoryEntry)
);

/**
 * getExampleSectionBreakdown(state)
 * returns section breakdown of first grades result.
 * @param {object} state - redux state
 * @return {object[]} - section breakdown of first grades result.
 */
export const getExampleSectionBreakdown = ({ grades }) => (
  (grades.results[0] || {}).section_breakdown || []
);

/**
 * gradeOverrides(state)
 * returns the gradeOverride history results
 * @param {object} state - redux state
 * @return {object[]} - grade override history result entries
 */
export const gradeOverrides = ({ grades }) => grades.gradeOverrideHistoryResults;

/**
 * hasOverrideErrors(state)
 * returns true iff gradeOverrideHistoryErrors are present
 * @param {object} state - redux state
 * @return {bool} - are there grade override history errors?
 */
export const hasOverrideErrors = ({ grades }) => !!grades.gradeOverrideHistoryError;

/**
 * uploadSuccess(state)
 * @param {object} state - redux state
 * @return {bool} - is bulkManagement.uploadSuccess true?
 */
export const uploadSuccess = ({ grades: { bulkManagement } }) => (
  !!bulkManagement && bulkManagement.uploadSuccess
);

const simpleSelectors = simpleSelectorFactory(
  ({ grades }) => grades,
  [
    'courseId',
    'filteredUsersCount',
    'totalUsersCount',
    'gradeFormat',
    'nextPage',
    'prevPage',
    'showSpinner',
    'gradeOverrideCurrentEarnedGradedOverride',
    'gradeOverrideHistoryError',
    'gradeOriginalEarnedGraded',
    'gradeOriginalPossibleGraded',
    'nextPage',
    'prevPage',
    'showSuccess',
  ],
);

export default StrictDict({
  bulkImportError,
  formatGradeOverrideForDisplay,
  formatMinAssignmentGrade,
  formatMaxAssignmentGrade,
  formatMaxCourseGrade,
  formatMinCourseGrade,
  hasOverrideErrors,
  headingMapper,

  roundGrade,
  subsectionGrade,

  ...simpleSelectors,
  allGrades,
  bulkManagementHistoryEntries,
  getExampleSectionBreakdown,
  gradeOverrides,
  uploadSuccess,
});
