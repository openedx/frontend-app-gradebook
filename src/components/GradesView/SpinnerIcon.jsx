/* eslint-disable react/sort-comp, react/button-has-type, import/no-named-as-default */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { Icon } from '@edx/paragon';

import selectors from 'data/selectors';

/**
 * <SpinnerIcon />
 * Simmple redux-connected icon component that shows a spinner overlay only if
 * redux state says it should.
 */
export const SpinnerIcon = ({ show }) => show && (
  <div className="spinner-overlay">
    <Icon className="fa fa-spinner fa-spin fa-5x color-black" />
  </div>
);
SpinnerIcon.defaultProps = {
  show: false,
};
SpinnerIcon.propTypes = {
  show: PropTypes.bool,
};

export const mapStateToProps = (state) => ({
  show: selectors.root.shouldShowSpinner(state),
});

export default connect(mapStateToProps)(SpinnerIcon);
