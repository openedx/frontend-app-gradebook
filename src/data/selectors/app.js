/* eslint-disable import/no-self-import */
import { StrictDict } from 'utils';
import simpleSelectorFactory from '../utils';
import * as module from './app';
import { minGrade, maxGrade } from './grades';

/**
 * assignmentGradeLimits(state)
 * returns an object of local assignmentGradeMin/Max
 * @param {object} state - redux state
 * @return {object} - { assignmentGradeMin, assignmentGradeMax }
 */
const assignmentGradeLimits = (
  { app: { filters: { assignmentGradeMin, assignmentGradeMax } } },
) => ({ assignmentGradeMin, assignmentGradeMax });

/**
 * courseGradeFilterValidity(state)
 * returns { isMaxValid, isMinValid } with each course grade limit, verifying they
 * are between minGrade and maxGrade (0, 100).
 * @param {object} state - redux state
 * @return {object} - { isMaxValid, isMinValid }
 */
export const courseGradeFilterValidity = ({ app: { filters } }) => {
  const isFilterValid = (value) => {
    const intValue = parseInt(value, 10);
    return intValue >= minGrade && intValue <= maxGrade;
  };
  return {
    isMaxValid: isFilterValid(filters.courseGradeMax),
    isMinValid: isFilterValid(filters.courseGradeMin),
  };
};

/**
 * courseGradeLimits(state)
 * returns an object of local courseGradeMin/Max
 * @param {object} state - redux state
 * @return {object} - { courseGradeMin, courseGradeMax }
 */
const courseGradeLimits = (
  { app: { filters: { courseGradeMin, courseGradeMax } } },
) => ({ courseGradeMin, courseGradeMax });

/**
 * editUpdateDate(state)
 * builds object for update api call from edit modal.
 * @param {object} state - redux state
 * @return {object} - object to pass as update date to updateGrades to apply editModal changes
 */
const editUpdateData = ({ app: { modalState } }) => ([{
  grade: {
    comment: modalState.reasonForChange,
    earned_graded_override: modalState.adjustedGradeValue,
  },
  usage_id: modalState.updateModuleId,
  user_id: modalState.updateUserId,
}]);

/**
 * areCourseGradeFiltersValid(state)
 * returns true iff both min and max course grade filters are valid (within bounds)
 * @param {object} state - redux state
 * @return {bool} - are both filters valid?
 */
const areCourseGradeFiltersValid = (state) => {
  const validity = module.courseGradeFilterValidity(state);
  return validity.isMinValid && validity.isMaxValid;
};

const isFilterMenuClosed = ({ app: { filterMenu } }) => (
  !filterMenu.open && !filterMenu.transitioning
);

const isFilterMenuOpening = ({ app: { filterMenu } }) => (
  filterMenu.transitioning && filterMenu.open
);

const modalSelectors = simpleSelectorFactory(
  ({ app: { modalState } }) => modalState,
  [
    'assignmentName',
    'adjustedGradePossible',
    'adjustedGradeValue',
    'open',
    'reasonForChange',
    'todaysDate',
    'updateUserName',
  ],
);

const filterMenuSelectors = simpleSelectorFactory(
  ({ app: { filterMenu } }) => filterMenu,
  ['open', 'transitioning'],
);

const simpleSelectors = simpleSelectorFactory(
  ({ app }) => app,
  [
    'activeView',
    'courseId',
    'filters',
    'searchValue',
    'showImportSuccessToast',
  ],
);

export default StrictDict({
  areCourseGradeFiltersValid,
  assignmentGradeLimits,
  courseGradeFilterValidity,
  courseGradeLimits,
  editUpdateData,
  isFilterMenuClosed,
  isFilterMenuOpening,
  ...simpleSelectors,
  modalState: StrictDict(modalSelectors),
  filterMenu: StrictDict({
    ...filterMenuSelectors,
    isClosed: isFilterMenuClosed,
    isOpening: isFilterMenuOpening,
  }),
});
