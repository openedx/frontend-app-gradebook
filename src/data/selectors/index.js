import { StrictDict } from 'utils';

import lms from 'data/services/lms';
import * as filterConstants from 'data/constants/filters';

import * as module from '.';
import app from './app';
import assignmentTypes from './assignmentTypes';
import cohorts from './cohorts';
import filters from './filters';
import grades, { minGrade, maxGrade } from './grades';
import roles from './roles';
import tracks from './tracks';

const {
  filterConfig,
  filters: filterNames,
} = filterConstants;

/**
 * editModalPossibleGrade(state)
 * Returns the "possible" grade as shown in the edit modal.
 * @param {object} state - redux state;
 * @return {string} - possibleGrade to show on edit modal
 */
export const editModalPossibleGrade = (state) => (
  app.modalState.adjustedGradePossible(state) || grades.gradeOriginalPossibleGraded(state)
);

/**
 * filterBadgeConfig(state, name)
 * Takes a filter name and returns the appropriate badge config, with value and isDefault.
 * Determines if it should return a range or single-value config based on the presence of
 * a filterOrder prop in the filter config associated with the passed name.
 * @param {object} state - redux state
 * @param {string} name - api filter name
 */
export const filterBadgeConfig = (state, name) => {
  const filterValue = module.filterBadgeValues[name](state);
  const { filterOrder, ...config } = filterConfig[name];
  const isRange = !!filterOrder;
  const value = isRange ? `${filterValue[0]} - ${filterValue[1]}` : filterValue;
  const isDefault = (isRange
    ? (
      filters.isDefault(filterOrder[0], filterValue[0])
      && filters.isDefault(filterOrder[1], filterValue[1])
    )
    : filters.isDefault(name, filterValue)
  );
  return { ...config, value, isDefault };
};

/**
 * filterBadgeValues methods
 * For each filter type with an associated badge, provides a selector that returns the
 * content of that badge
 */
export const filterBadgeValues = StrictDict({
  [filterNames.assignment]: (state) => (
    filters.selectedAssignmentLabel(state) || ''
  ),
  [filterNames.assignmentType]: filters.assignmentType,
  [filterNames.includeCourseRoleMembers]: filters.includeCourseRoleMembers,
  [filterNames.cohort]: (state) => {
    const entry = module.selectedCohortEntry(state);
    return entry ? entry.name : '';
  },
  [filterNames.track]: (state) => {
    const entry = module.selectedTrackEntry(state);
    return entry ? entry.name : '';
  },
  [filterNames.assignmentGrade]: (state) => ([
    filters.assignmentGradeMin(state),
    filters.assignmentGradeMax(state),
  ]),
  [filterNames.courseGrade]: (state) => ([
    filters.courseGradeMin(state),
    filters.courseGradeMax(state),
  ]),
});

/**
 * formattedGradeLimits(state)
 * Returns an object of local grade limits, formatted for fetching.
 * This means only setting any of them if the assignment is set, and only setting an
 * individual min/max if they are not equal to the default min/max.
 * @param {object} state - redux state
 * @return {object} - { assignmentGradeMax, assignmentGradeMin, courseGradeMax, courseGradeMin }
 */
export const formattedGradeLimits = (state) => {
  const { assignmentGradeMax, assignmentGradeMin } = app.assignmentGradeLimits(state);
  const { courseGradeMax, courseGradeMin } = app.courseGradeLimits(state);
  const hasAssignment = filters.selectedAssignmentId(state) !== undefined;

  return {
    assignmentGradeMax: (assignmentGradeMax === maxGrade || !hasAssignment) ? null : assignmentGradeMax,
    assignmentGradeMin: (assignmentGradeMin === minGrade || !hasAssignment) ? null : assignmentGradeMin,
    courseGradeMax: courseGradeMax === maxGrade ? null : courseGradeMax,
    courseGradeMin: courseGradeMin === minGrade ? null : courseGradeMin,
  };
};

/**
 * getHeadings(state)
 * Returns the table headings given the current assignmentType and assignmentLabel filters.
 * @param {object} state - redux state
 * @return {string[]} - array of table headings
 */
export const getHeadings = (state) => grades.headingMapper(
  filters.assignmentType(state) || 'All',
  filters.selectedAssignmentLabel(state) || 'All',
)(grades.getExampleSectionBreakdown(state));

/**
 * gradeExportUrl(state, options)
 * Returns the output of getGradeCsvUrl, applying the current includeCourseRoleMembers
 * filter.
 * @param {object} state - redux state
 * @return {string} - generated grade export url
 */
export const gradeExportUrl = (state) => (
  lms.urls.gradeCsvUrl(module.lmsApiServiceArgs(state))
);

/**
 * interventionExportUrl(state, options)
 * Returns the output of getInterventionExportUrl.
 * @param {object} state - redux state
 * @return {string} - generated intervention export url
 */
export const interventionExportUrl = (state) => (
  lms.urls.interventionExportCsvUrl(module.lmsApiServiceArgs(state))
);

/**
 * lmsApiServiceArgs(state)
 * Returns common lms api service request args.
 * @param {object} state - redux state
 * @return {object} lms api query params object
 */
export const lmsApiServiceArgs = (state) => ({
  cohort: cohorts.getCohortNameById(state, filters.cohort(state)),
  track: filters.track(state),
  assignment: filters.selectedAssignmentId(state),
  assignmentType: filters.assignmentType(state),
  assignmentGradeMin: grades.formatMinAssignmentGrade(
    filters.assignmentGradeMin(state),
    { assignmentId: filters.selectedAssignmentId(state) },
  ),
  assignmentGradeMax: grades.formatMaxAssignmentGrade(
    filters.assignmentGradeMax(state),
    { assignmentId: filters.selectedAssignmentId(state) },
  ),
  courseGradeMin: grades.formatMinCourseGrade(filters.courseGradeMin(state)),
  courseGradeMax: grades.formatMaxCourseGrade(filters.courseGradeMax(state)),
  excludedCourseRoles: filters.excludedCourseRoles(state),
});

/**
 * localFilters(state)
 * returns local filter data for fetchGrades call
 * @param {object} state - redux state
 * @return {object} - fetch arguments signifying current local filter state
 */
export const localFilters = (state) => {
  const id = filters.selectedAssignmentId(state);
  const searchText = app.searchValue(state);
  return {
    assignment: id,
    includeCourseRoleMembers: filters.includeCourseRoleMembers(state),
    ...module.formattedGradeLimits(state),
    ...(searchText !== '' && { searchText }),
  };
};

/**
 * selectedCohortEntry(state)
 * Returns the full entry data for the selected cohort
 * @param {object} state - redux state
 * @return {object} - selected cohort entry object
 */
export const selectedCohortEntry = (state) => (
  cohorts.allCohorts(state).find(
    ({ id }) => id === parseInt(filters.cohort(state), 10),
  )
);

/**
 * selectedTrackEntry(state)
 * Returns the full entry data for the selected track
 * @param {object} state - redux state
 * @return {object} - selected track entry object
 */
export const selectedTrackEntry = (state) => (
  tracks.allTracks(state).find(
    ({ slug }) => slug === filters.track(state),
  )
);

/**
 * shouldShowSpinner(state)
 * Returns true iff the user can view the gradebook and grades.showSpinner is true.
 * @param {object} state - redux state
 * @return {bool} - should show spinner?
 */
export const shouldShowSpinner = (state) => (
  roles.canUserViewGradebook(state)
  && grades.showSpinner(state)
);

/**
 * showBulkManagement(state, options)
 * Returns true iff the course has bulk management enabled
 * @param {object} state - redux state
 * @return {bool} - should show bulk management controls?
 */
export const showBulkManagement = (state) => (state.config.bulkManagementAvailable);

export default StrictDict({
  root: StrictDict({
    editModalPossibleGrade,
    filterBadgeConfig,
    getHeadings,
    gradeExportUrl,
    interventionExportUrl,
    localFilters,
    selectedCohortEntry,
    selectedTrackEntry,
    shouldShowSpinner,
    showBulkManagement,
  }),
  app,
  assignmentTypes,
  cohorts,
  filters,
  grades,
  roles,
  tracks,
});
