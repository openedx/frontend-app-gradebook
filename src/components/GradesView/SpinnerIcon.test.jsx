import React from 'react';
import { shallow } from 'enzyme';

import selectors from 'data/selectors';
import { SpinnerIcon, mapStateToProps } from './SpinnerIcon';

jest.mock('@edx/paragon', () => ({
  Icon: () => 'Icon',
}));
jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    root: { shouldShowSpinner: state => ({ shouldShowSpinner: state }) },
  },
}));

describe('SpinnerIcon', () => {
  describe('component', () => {
    it('snapshot - does not render if show: false', () => {
      expect(shallow(<SpinnerIcon />)).toMatchSnapshot();
    });
    test('snapshot - displays spinner overlay with spinner icon', () => {
      expect(shallow(<SpinnerIcon show />)).toMatchSnapshot();
    });
  });
  describe('mapStateToProps', () => {
    const testState = { a: 'nice', day: 'for', some: 'sun' };
    test('show from root.shouldShowSpinner', () => {
      expect(mapStateToProps(testState).show).toEqual(selectors.root.shouldShowSpinner(testState));
    });
  });
});
