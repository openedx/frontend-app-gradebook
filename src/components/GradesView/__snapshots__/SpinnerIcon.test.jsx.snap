// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinnerIcon component snapshot - displays spinner overlay with spinner icon 1`] = `
<div
  className="spinner-overlay"
>
  <Icon
    className="fa fa-spinner fa-spin fa-5x color-black"
  />
</div>
`;

exports[`SpinnerIcon component snapshot - does not render if show: false 1`] = `""`;
